JSON-LD (a.k.a. RDF) vocabulary and SHACL shape for the DID Core Vocabulary

iherman opened the pull request:
This is an initial version for a proper RDF vocabulary, as well as a SHACL shape, for DID Documents using the JSON-LD serialization _and_ that may also be used as part of a larger Linked Data environment. This PR is related to https://github.com/w3c/did-core/issues/404 (and also to https://github.com/w3c/did-core/pull/536).

(The shape document and the RDF vocabulary use the Turtle syntax, which is more appropriate for a succinct definition of RDF triples. At some point, when both are final, it is worth producing a JSON-LD version of both.)

The PR contains a number of open issues, subject to further discussions; the PR should _not_ be merged until those are discussed and resolved.

## Vocabulary issues

_At the moment_ most of the DID core vocabulary items are defined in the [CCG Security Vocabulary](https://w3id.org/security#) and its corresponding namespace: `https://w3id.org/security#`. Note, however, that there is no formal (RDF) vocabulary definition for the security vocabulary, which breaks the expectation on the Linked Data to be able to “follow your nose”, i.e., being able to find out some more data about a term in machine readable form by following the term’s URL. (The vocabulary file attempts to do that for the DID Core terms).

 This raises some issues to discuss:

- **Where is the authoritative definition of a DID Core term?** in my view, this should be the DID Core specification. Note that the Security Vocabulary document may include some definition for a term used by DID (see, e.g., [verificationMethod](https://w3c-ccg.github.io/security-vocab/#verificationMethod)) but that definition is usually more shallow than what is in the DID Core document, and is also often labeled as having an 'unstable' status.

    My proposal would be that for _all_ DID Core terms the CCG Security Vocabulary would defer to the DID Core specification (i.e., the vocabulary text would just include a pointer to the DID spec). (Note that this is what the AS namespace document does for `as:alsoKnownAs`.)

- **What is the exact name of some vocabulary items?** Looking at the [DID JSON-LD context file](https://github.com/w3c/did-spec-registries/blob/master/contexts/did-v1.jsonld) the term `capabilityInvocation`, for example, maps on `sec:capabilityInvocationMethod`. On the other hand, the CCG Security Vocabulary document (and the examples therein) only refers to `capabilityInvocation`. This discrepancy is a bug somewhere; it occurs for `authentication`, `keyAgreement`, `capabilityDelegation`, and `capabilityInvocation`.

    At the moment, the vocabulary document in this PR follows the context file mapping.

- **What about the `did` namespace?** At the moment, that namespace only includes the `service` and `serviceEndpoint` terms (the vocabulary in the PR adds some classes for technical reasons). Technically this is correct, and the vocabulary file may describe the various information referring the security namespace. However, this is a bit touchy and “smells” some level of vocabulary hijacking: a formal definition of a term in a different namespace being part of the `did` namespace document is, shall we say, unusual. Also, honestly, there is an element of surprise for the usage of `sec` terms, defined formally in the DID Core document and not being in the `did` namespace…

   I presume the fact of using the `sec` namespace cannot be changed anymore. A (controversial) approach I have taken therefore in the vocabulary is:

   1. Define _all_ terms in the `did` namespace, including the range and domain statements
   2. Add a `owl:sameAs` on all terms, when appropriate, binding it to its `sec` counterpart. (Note that this trick is even more important if the vocabulary name discrepancy for some property names remains.)

- **What is the value of `sec:publicKeyJwk`?** This is less controversial: the DID context file defines the value type of this property with the `@json` keyword. The vocabulary expands on that, making use of the fact that `@json` formally maps on the `rdf:JSON` datatype.

Note that, at least for the time being, the vocabulary is very simple and uses, mostly, RDFS terms only. If we agree with the overall direction, I am happy to expand it with a richer set of statements that would then also rely on OWL (e.g., by defining the domain of the `controller` property as a union of classes).


## SHACL issues

I think I could cover most of the constraints in the document using SHACL. Except one: I do not see any way the _restrict_ an identifier (i.e., a URL) to be of a specific format (i.e., a DID). I am afraid that would go beyond the RDF model…

OR13 commented on vocabs/DID-core-shape.ttl:
```diff
@@ -0,0 +1,129 @@
+#  pyshacl -s DID-core-shape.ttl -e DID-core.ttl -f human test.ttl
+#
+
+@prefix did:  <https://w3id.org/did#> .
```
are we sure w3id.org is the right choice for this prefix? 

it currently redirect without preserving fragment... https://w3id.org/did#dfn-alsoknownas

I would have thought this would point to vocabulary file, like:

```
@prefix as:   <https://www.w3.org/ns/activitystreams> .
```

does.

iherman replied:
This is what is in the `@context` file... (a.k.a. don't kill the messenger :-)

OR13 replied:
killing messengers results in better messages ;)

@rhiaro @msporny I won't accept this PR as is :)

TallTed replied:
@OR13 -- I would say "yes, w3id.org is the right choice". You set up the w3id.org redirects for `did-resolution`.  You should know that fixing this issue of the redirect-without-fragment (which might just be to make the redirect target retain the fragment, as in https://w3c.github.io/did-core/#dfn-alsoknownas) just requires an issue and/or PR against <https://github.com/perma-id/w3id.org/blob/master/did/.htaccess>. PR best to come from a past contributor to that (who might also add a README to describe the DID redirect and indicate who may make future changes to the `did` folder of the https://github.com/perma-id/w3id.org/ repo), i.e., @dlongley, @msporny, or @davidlehn.

rhiaro replied:
@iherman Which context file are you looking at? The [current v1 here](https://github.com/w3c/did-spec-registries/blob/master/contexts/did-v1.jsonld) does not use w3id.org at all

iherman replied:
Hm. You are right (maybe I looked at an older file, I am not sure). But there is a problem nevertheless. The `@context` file says:.

```
"didns": "https://www.w3.org/ns/did/v1#"
```

but if I dereference that URL I get to the `@context` file. That is incorrect. A JSON-LD `@context` file is _not_ a vocabulary file. What I expect to see at that URL is the RDFS definition of the terms, i.e., what I have put into this PR.

rhiaro replied:
@iherman yes, I think we need a separate conversation about putting the right files under the right places at /ns/ and sorting out the conneg. We also have a html namespace file now that needs to go there at some point.

iherman replied:
> Which context file are you looking at? The current v1 here does not use w3id.org at all

B.t.w., I just realized that the context file has been changed since I branched off... hence the discrepancy.

OR13 commented on vocabs/DID-core-shape.ttl:
```diff
@@ -0,0 +1,129 @@
+
+    sh:targetClass did:DIDDocument ;
+    sh:targetSubjectsOf
+        did:alsoKnownAs,
```
consider adding canonicalId / equivalentId here cc @csuwildcat

iherman replied:
Those are not in the core vocabulary of the DID specification, they are resolution metadata properties. *If* we introduced them in the core vocabulary, that means they could be used in a DID Document, and I believe the WG firmly voted agaist that (sorry @csuwildcat...)

csuwildcat replied:
I'm not sure what this means, but I hope it doesn't mean the properties are being removed from the spec, still have the same gravity, and include the same normative MUST statements.

iherman replied:
@csuwildcat 

> I'm not sure what this means, but I hope it doesn't mean the properties are being removed from the spec, still have the same gravity, and include the same normative MUST statements.

To put your mind at ease: no, it doesn't.

The vocabulary, as well as the SHACL constraint language, is only relevant to the [core terms in the model](https://w3c.github.io/did-core/#core-properties). To be even more precise: the vocabulary is of relevance if and only if the DID Document is expressed in JSON-LD and the RDF mapping of that JSON-LD is of use by a specific application (method or otherwise). All this completely orthogonal to the resolution metadata properties, their specification, and usage.

OR13 commented on vocabs/DID-core-shape.ttl:
```diff
@@ -0,0 +1,129 @@
+        [
+            sh:path did:publicKeyJwk ;
+            sh:nodeKind sh:Literal ;
+            sh:datatype rdf:JSON ;
```
<3

OR13 commented on vocabs/DID-core.ttl:
```diff
@@ -0,0 +1,122 @@
+    rdfs:range rdf:JSON ;
+    rdfs:isDefinedBy <https://www.w3.org/TR/did-core/#verification-methods> ;
+    rdfs:label "Public Key in JWK" ;
+    owl:sameAs sec:publicKeyJwk ;
```
<3

OR13 reviewed (changes requested):
I think the did prefix should behave the same way as the as prefix, we should open an issue to track fixing that, or address it before merging this.

iherman commented:
> I think the did prefix should behave the same way as the as prefix, we should open an issue to track fixing that, or address it before merging this.

As I said, I have just copied what is in the `@context` file. I am in favour of using the `www.w3.org/ns` for this vocabulary URI, but that is my personal opinion, with my W3C staff contact's hat put _down_.

OR13 commented:
@iherman I am also in favor of using `www.w3.org/ns` instead of `w3id.org`, and am requesting that change here :)

The rest of this PR looks great! thanks for tackling this.

msporny commented:
> @iherman I am also in favor of using `www.w3.org/ns` instead of `w3id.org`, and am requesting that change here :)

I'm fairly certain that should be fine... /cc @dlongley @dmitrizagidulin who may have stronger opinions about RDF DID vocabulary namespace. Have we shipped anything into production that depends on the "old" w3id.org DID namespace?

dlongley commented:
> Have we shipped anything into production that depends on the "old" w3id.org DID namespace?

Not to my knowledge. I'm ok with the change. To be clear, there are only a few terms that are in the DID namespace, mostly around services.

iherman commented:
As I said in https://github.com/w3c/did-spec-registries/pull/182#issuecomment-758578461 we should have a consistent URI scheme, though. A first proposal would be:

1. JSON-LD `@context` file: `https://www.w3.org/ns/did/v1.jsonld`
2. DID RDF vocabulary space: `https://www.w3.org/ns/did/vocab`, use content negotiation to turtle, jsonld, possibly to rdf/xml 
3. DID RDF vocabulary shape: `https://www.w3.org/ns/did/shacl`, use content negotiation to turtle, jsonld
4. DID JSON Schema: `https://www.w3.org/ns/did/schema`

If we were just starting I would have preferred to use `https://www.w3.org/ns/did/context` for the context file. But that ship has sailed.

Happy bike shedding :-)

TallTed commented on vocabs/readme.md:
```diff
@@ -0,0 +1,13 @@
+
+## Notes
+
+* The URI-s used for the terms and the vocabulary as a whole may still change.
```
```suggestion
* The URIs used for the terms and the vocabulary as a whole may still change.
```

OR13 reviewed (changes requested):
I really want to take these changes, but I would prefer

https://w3id.org/did#alsoKnownAs

resolve somewhere before accepting them.... 2 options....

1. update w3id.org to redirect to w3.org/TR https://github.com/perma-id/w3id.org/blob/master/did/.htaccess

2. remove w3id.org from the shacl vocab.

msporny commented:
> update w3id.org to redirect to w3.org/TR https://github.com/perma-id/w3id.org/blob/master/did/.htaccess

+1 to doing the above -- doesn't it already -- maybe you mean point to TR space instead of the Editor's Draft? Or is the issue that the fragment isn't preserved or something like that?

rhiaro commented on vocabs/DID-core.jsonld:
```diff
@@ -0,0 +1,144 @@
+      "@id" : "https://w3id.org/did#",
+      "@type" : "owl:Ontology"
+    }, {
+      "@id" : "did:DIDDocument",
```
Sorry I'm late to review this. I'm a bit concerned about adding these classes. Especially the blurring of the DID Subject and DID Document here. Maybe it is completely harmless, but the RDF users in my experience would expect to see such terms documented in the spec or namespace document, and they won't be (and adding them is going to confuse folks who aren't interested in the RDF side of this, I'm sure).

I'm also not sure about the implications on inferencing of having these classes. We're not going to leave the door open for incorrect assertions to be made here, are we? You have probably thought all of this through though.

rhiaro replied:
By which I mean, by using `did:DIDDocument` as the domain of various properties, we are implying that the property values apply to the DID document, and not the DID subject, are we not? But I thought there was general agreement that there is 'no such thing' as a DID document, rather it is a serialisation of attributes which apply to the DID subject (which in some cases is a document, and in some cases is a person or a cat, etc). I don't want to re-open that debate though.

iherman replied:
Removing the classes from the vocabulary means

- We cannot properly define property domains and ranges, meaning that the vocabulary file would 'just' be a series of `rdf:property` statements. We can do that, but it becomes a bit meaningless. We can document the classes in the vocabulary itself, and making it clear why it is not in the formal spec... 

    Fair to say that I do not have the good background to judge whether, from a security/privacy point of view, there is any harm in these; I do not see any. @dlongley or @msporny should say that. But, as a general notion, if we do use JSON-LD, ie, we use RDF(S), then the open world assumption stands for DID Documents, too...

- I am not sure the SHACL would work correctly. The "shapes" are attached to classes (SHACL makes use of the vocabulary). At least the 'starting' one, namely DIDDocument class, is needed (ShEx does not have this problem because it expects an explicit "mapping")

iherman replied:
> By which I mean, by using did:DIDDocument as the domain of various properties, we are implying that the property values apply to the DID document, and not the DID subject, are we not?

Aha! Maybe there is a naming thing here, which is my mistake.

As you may see there is no such concept in the vocabulary as DID subject because that is 'just' the URL of the, well, subject. Which makes me realize that what I called the `DIDDocument` class should be named `DIDSubject`!

Ie, it should say:

```
did:verificationMethod a rdf:Property ;
    rdfs:range did:VerificationMethod ;
    rdfs:domain did:DIDSubject ;
    ....
```

which means that if I have data of the form

```
<did:ex:123456> did:verificationMethod <did:ex:5t5t5t5t#something>
```

Then the inference would say that `<did:ex:123456>` is of class `DIDSubject`. And this is what we want.

Bottom line: I made a mistake in naming.

rhiaro commented:
> update w3id.org to redirect to w3.org/TR

We want to update w3id.org to redirect w3.org/ns/did[/v1?] right? And get the HTML namespace document that has been in this repo for a while uploaded there?

iherman commented:
> > update w3id.org to redirect to w3.org/TR
> 
> We want to update w3id.org to redirect w3.org/ns/did[/v1?] right? And get the HTML namespace document that has been in this repo for a while uploaded there?

What about the other URL-s (https://github.com/w3c/did-spec-registries/pull/183#issuecomment-759232149) ?

rhiaro commented:
> What about the other URL-s (#183 (comment)) ?

We probably want to insert `/v1/` but otherwise I have no opinion on the shacl or the schema URIs. 

I understand that rdfs vocabs and jsonld contexts are different, but I thought it was common practice to return the jsonld context with ld+json is requested from the namespace URI. Is that incorrect?

TallTed commented:
We do want the w3id.org redirects to work properly, including fragment IDs.  For the immediate, while all these docs are still in draft phase, these redirects should (continue to) go to the latest draft of the relevant resource, generally in github.io space.  When (soon) the drafts are promoted to CR or later, the w3id.org redirects should be, uh, redirected to go to the relevant resource in `w3.org/ns/` or `w3.org/TR/` or whatever space, with `/v1/` or whatever we've decided is appropriate.  Nothing terribly complicated nor unusual in this...

iherman commented:
> 
> I understand that rdfs vocabs and jsonld contexts are different, but I thought it was common practice to return the jsonld context with ld+json is requested from the namespace URI. Is that incorrect?

I believe it is. The vocabulary can be (and is in this case) serialized in JSON-LD as well; if I ask for a vocabulary than that is what I should get and not the context.

iherman commented:
> We probably want to insert /v1/ but otherwise I have no opinion on the shacl or the schema URIs.

Agree with `/v1/`. I will prepare a new commit with all these changes so that we can move on.

rhiaro commented:
> I believe it is. The vocabulary can be (and is in this case) serialized in JSON-LD as well; if I ask for a vocabulary than that is what I should get and not the context.

So the problem is we have used the same URI for the context value and for the namespace? And it's probably too late to change either.

Since the spec explicitly supports JSON-LD but doesn't mention any other RDF serializations, I think it is important to return the JSON-LD context from that URI and hold our nose about the JSON-LD serialisation of the vocab (it is still there in turtle after all, anyone can convert it). Not to mention, JSON-LD processors would require that to work properly, no?

iherman commented:
I have made the following changes:

* The URL-s have been changed in the files such that:
    * `did` is now set to `https://www.w3.org/ns/did/vocab/v1#`
    * the shex file is to be reached via `https://www.w3.org/ns/did/shex/v1`
    * the shacl file is to be reached via `https://www.w3.org/ns/did/shacl/v1`

    The files in the repository have also been renamed by adding 'v1' to the names. 
* The `DIDDocument` class has been renamed to `DIDObject`
* Removed the domain settings for `verificationMethod` and friends, because they _may_ be used on other instances (e.g., on services, although it is not part of the spec but was raised as a possibility) and the inference to `DIDObject` should not happen in those cases
* The value of `publicKeyJwk` is a literal of type `rdf:JSON` (that was actually done in an earlier commit, but may not have been properly reported here)

The reason I used `https://www.w3.org/ns/did/vocab/v1#` and not `https://www.w3.org/ns/did/v1/vocab#` is not to mess up with the current redirections... (at the moment, 'v1' behaves like file name and not as a folder name).

---

> So the problem is we have used the same URI for the context value and for the namespace? And it's probably too late to change either.

I do not see why it would be a problem to change that. Current implementations use the context file, they do not use the URI of the DID namespace directly. As it seems the DID namespace URI has been changed in the context just a few days ago (see https://github.com/w3c/did-spec-registries/pull/183#discussion_r563130314) and nobody even noticed.

_Note to myself:_ if this branch is merged then the context file must be updated to use `https://www.w3.org/ns/did/vocab/v1#`

> Since the spec explicitly supports JSON-LD but doesn't mention any other RDF serializations, I think it is important to return the JSON-LD context from that URI and hold our nose about the JSON-LD serialisation of the vocab 

What we have is:
- The widely documented and used URI for the context, ie, `https://www.w3.org/ns/did/v1` returns the context file in jsonld. That has not changed. 
- The vocab can be reached via the `https://www.w3.org/ns/did/vocab/v1`; by default it returns the vocabulary in turtle, but has an access to jsonld and rdf/xml with conneg. 
- `https://www.w3.org/ns/did/shacl/v1` returns the SHACL file, default to turtle, and conneg to jsonld
- `https://www.w3.org/ns/did/shex/v1` returns the shex file in SHEXC syntax (that is the only syntax). 
 
I believe this should work and cover all the needs.

(The redirections have been set up for vocab and the shapes but become fully operational once the branch is merged.)

_Note to myself:_ a similar setup for `https://www.w3.org/ns/did/schema/v1` should be done if this branch is merged.

@msporny @dlongley @OR13

msporny commented:
> > We probably want to insert /v1/ but otherwise I have no opinion on the shacl or the schema URIs.
> 
> Agree with `/v1/`. I will prepare a new commit with all these changes so that we can move on.

Hold up, wait a second! :)

Just want to make sure we're all on the same page: 

`.../security/v1` -> should always result in the JSON-LD Context... never the vocabulary file
`.../security/` -> should always result in the human readable RDF Vocabulary by default.

You can conneg for other representations than the default, but each representation should be different serializations of the same basic content as the above.

There is no such thing as `.../security/v1/` <-- that should redirect you to `.../security/v1`

We use this design pattern because:

1. Developers tend to get the conneg stuff wrong, so the defaults should align with the 80%+ case.
2. Developers have stated a preference to not have versioned RDF Vocabularies and having one place to look for all vocabulary terms across all versions.
3. The pattern allows us to do breaking changes with major version releases (e.g., v1 to v2 to v3) while not breaking the already deployed ecosystem.

Does that make sense?

msporny commented:
> The widely documented and used URI for the context, ie, https://www.w3.org/ns/did/v1 returns the context file in jsonld. That has not changed.

+1

> The vocab can be reached via the https://www.w3.org/ns/did/vocab/v1; by default it returns the vocabulary in turtle, but has an access to jsonld and rdf/xml with conneg.

-1, the conneg vocab should be at `https://www.w3.org/ns/did` (with a default to text/html). All of our DID vocab URLs should be of the form `https://www.w3.org/ns/did#exampleVocabTerm`.

> https://www.w3.org/ns/did/shacl/v1 returns the SHACL file, default to turtle, and conneg to jsonld
> https://www.w3.org/ns/did/shex/v1 returns the shex file in SHEXC syntax (that is the only syntax).

Why can't we conneg for that? with the default for `application/json` and `application/ld+json` being the JSON-LD Context? Or at the very least, with a file extension? `https://www.w3.org/ns/did/v1.shex` and `https://www.w3.org/ns/did/v1.shacl`?

msporny commented:
Gah, sorry @iherman -- my comments were 20 minutes too late :(

rhiaro commented:
> -1, the conneg vocab should be at https://www.w3.org/ns/did (with a default to text/html). All of our DID vocab URLs should be of the form https://www.w3.org/ns/did#exampleVocabTerm. We want it to be easy to type for developers and to line up with the URLs in the JSON-LD Context.

@msporny Right now the JSON-LD Context URI is *the same* as the namespace URI, ie. `https://www.w3.org/ns/did/v1#"`

Are you saying the JSON-LD context should be `https://www.w3.org/ns/did/v1` but the namespace URI *in* the JSON-LD context should be `https://www.w3.org/ns/did#`? I'll PR to fix that if so. And for the human-readable content at `/ns/did` can be the same as at `/ns/did/v1` right now (which is ready in this repo), and always just direct to the latest version of the human-readable vocab?

iherman commented:
@msporny, I cannot test things right now (I am not at my machine any more) but, while I see where you are going, I am not sure I can do it (technically).

At the moment, did is a directory under /ns, it uses a v1.var file plus a .htaccess encoded redirection towards the repository. That works; v1.var makes conneg smooth.

I am not sure how to set up the conneg for /ns/did without harming what is there. I could of course det up a /ns/did.var file and all the redirections on that level, but that would harm the current setup. I could create a /ns/did/vocab.var and the we'd have 'https://www.w3.org/did/vocab#...' for the terms, which is very close to what I did, but doesn't have the 'v1' in it, which isn't consistent with the way the context is named...

As for shacl: while shex has a different media type for its shexc syntax, shacl doesn't: it is simply rdf. Ie, it would interfere with other conneg if it was on the same urls as the vocabulary...

I may miss some obvious setup trick, though...

msporny commented:
@rhiaro wrote:
> Are you saying the JSON-LD context should be https://www.w3.org/ns/did/v1 but the namespace URI in the JSON-LD context should be https://www.w3.org/ns/did#? 

Yes, correct.

> I am not sure I can do it (technically).

I know we've been able to achieve this pattern using both apache and nginx .htaccess files before. I'm fairly certain W3C infrastructure can do this... it might take some trial and error.

We can always use file extensions to achieve what we want (and should)... so, from a physical files perspective... there should be `did/index.html`, `did/v1.jsonld`, `did/v1.shex`, `did/v1.shacl`... and then conneg entries setup for each as well... and maybe since SHACL doesn't have it's own media type, and that community probably wouldn't be cranky about the extension, you can't conneg for it.

iherman commented:
@msporny,

I would like to understand what you expect, because it is not clear.

Is it so that by dereferencing `https://www.w3.org/ns/did/v1` you'd expect:

1. when asking for text/html (or by default): an HTML file describing the vocabulary;
2. when asking for application/ld+json: the `@context` file;
3. when asking for text/turtle: the RDFS vocabulary file;
4. when asking for text/shex: the SHEX file;

If this is what you are asking for, a big -1 for me. Indeed, that would mean that, depending on the media type you are asking for, the content would be different: a vocabulary file is different from a JSON-LD context file and is again different from a shex file. That is not what conneg should mean. Also, an RDF processor (or a SHACL processor!), when trying to dereference the vocabulary URL (e.g., for some level of reasoning), and asking for a JSON-LD serialization of the RDFS data, will get the context file which, on that level, is irrelevant for the caller and will break the reasoning or constraint engine. (Let alone the fact that there is no way to get the shacl in this list because it does not have a separate media type and reserved suffix.)

I believe this is an anti-pattern. 

If we want the 'v1' in the denomination of the vocabulary and shape files, then I stay by my proposal in https://github.com/w3c/did-spec-registries/pull/183#issuecomment-766094311. If adding the v1 to the vocabulary file is not really important then we can set up this:

1. `https://www.w3.org/ns/did/v1` refers to the context file (this is something we should not change anymore, I presume)
2. `https://www.w3.org/ns/did/vocab#` refers to the vocabulary file with conneg for turtle, json-ld, and rdf/xml (and this is the namespace URL for the vocabulary)
3. `https://www.w3.org/ns/did/shacl` refers to the SHACL with conneg for turtle and json-ld
4. `https://www.w3.org/ns/did/shex` refers to the SHEX in shexc syntax (we could also store shexj if I find a proper translator, but that is a detail)

For each case we can have an index.html for the HTML case but that is only a human-readable description of the real content; a side-issue, thus.

OR13 commented:
@msporny please approve this PR or request changes.

msporny commented:
> @msporny please approve this PR or request changes, this PR is blocking #237 as well.

It's on my review list for this weekend -- it's a long and complex PR and I wanted to make sure to do a good review. If I haven't done my review by next Monday, feel free to merge w/o my input.

rhiaro commented:
> The DIDDocument class has been renamed to DIDObject

Wait wait why another term? What happened to `DIDSubject`?

rhiaro commented:
> 1. https://www.w3.org/ns/did/v1 refers to the context file (this is something we should not change anymore, I presume)
> 2. https://www.w3.org/ns/did/vocab# refers to the vocabulary file with conneg for turtle, json-ld, and rdf/xml (and this is the namespace URL for the vocabulary)
> 3. https://www.w3.org/ns/did/shacl refers to the SHACL with conneg for turtle and json-ld
> 4. https://www.w3.org/ns/did/shex refers to the SHEX in shexc syntax (we could also store shexj if I find a proper translator, but that is a detail)
> 
> For each case we can have an index.html for the HTML case but that is only a human-readable description of the real content; a side-issue, thus.

That looks right to me except can 2. be `https://www.w3.org/ns/did#` instead? Is that really impossible with the way the w3c servers work?

iherman commented:
> > The DIDDocument class has been renamed to DIDObject
> 
> Wait wait why another term? What happened to `DIDSubject`?

It happened that yours truly is stupid and should be fired :-(

I changed DIDObject to DIDSubject.

(It is great to have @rhiaro a.k.a. eagle-eye around:-)

iherman commented:
> 
> That looks right to me except can 2. be `https://www.w3.org/ns/did#` instead? Is that really impossible with the way the w3c servers work?

I am not really an apache expert (and I do not think the W3C server setup is very different from other places)...

Here is what we have now:

- `/ns/did/` is a real directory on our sever
- `/ns/did/v1.var` takes care of conneg [1]
- `/ns/did/.htaccess` takes care of the redirections to github [2]

we know this works, and the mechanism can be reproduced easily to the two scenarios I mentioned (with or without the v1).

Introducing `/ns/did#` into the mix would require to add a `did.var` on the `/ns/` level, and add rules to `/ns/.htaccess`. I have the right to do these, so that is not the problem. However, what I do not know is whether the introduction of a `v1.var` would clash or not with the existence of a `/ns/did/` as a directory. My usual adage is to say "I know the jungle and therefore I am afraid of the jungle", meaning that I know fussing around with Apache setting is usually full of unintended consequences. Hence my somewhat simpler proposal (which, I think, would not have any real influence on deployment, it is only a matter of aesthetics.)

I am happy to experiment with it somewhere on a separate area on the W3C website, if it is really necessary. But is it?

(What may be easier is to set up is `https://www.w3.org/ns/did/`, but that would mean a separate file for each term in the case of a vocabulary. I think that would be really more complicated to maintain, hence the usual usage of `#` for vocabularies, so I would prefer not to do that.)





----------
[1] the `v1.var` looks like this:

```
URI: v1

URI: v1.jsonld
Content-Type: application/ld+json; qs=0.5

URI: v1.json
Content-Type: application/json; qs=0.4
```

[2] the `.htccess` looks like this:

```
RewriteEngine On
RewriteBase /ns/did/
AddType application/ld+json .jsonld


# Changes added on the latest block due URL name changes, 2019-10-12 (Ivan Herman)
RewriteRule ^v1.jsonld$ https://w3c.github.io/did-spec-registries/contexts/did-v1.jsonld [P]
RewriteRule ^v1.json$ https://w3c.github.io/did-spec-registries/contexts/did-v1.jsonld [P]
```

msporny commented on vocabs/DID-core-shape-v1.jsonld:
```diff
@@ -0,0 +1,228 @@
+            "@id": "_:b12",
+            "@type": "sh:NodeShape",
+            "property": [
+                "_:b10",
+                "_:b13",
+                "_:b14",
+                "_:b1",
+                "_:b9",
+                "_:b0",
+                "_:b15",
+                "_:b16",
+                "_:b17"
```
Can we use more descriptive blank node names here? For example _:controller, _:serviceEndpoint, etc.

iherman replied:
Well... I write the shacl (and the vocabulary) in Turtle and I let some tool translate it into JSON-LD. Once everything is final I can try to do some manual rewrite of these names, but not until we are absolutely final.

msporny commented on vocabs/DID-core-shape-v1.jsonld:
```diff
@@ -0,0 +1,228 @@
+        {
+            "@id": "_:b13",
+            "nodeKind": "sh:IRI",
+            "path": "did:verificationMethod"
```
```suggestion
            "path": "sec:verificationMethod"
```

msporny commented on vocabs/DID-core-shape-v1.jsonld:
```diff
@@ -0,0 +1,228 @@
+        {
+            "@id": "_:b14",
+            "nodeKind": "sh:IRI",
+            "path": "did:authentication"
```
```suggestion
            "path": "sec:authentication"
```

msporny commented on vocabs/DID-core-shape-v1.jsonld:
```diff
@@ -0,0 +1,228 @@
+        {
+            "@id": "_:b15",
+            "nodeKind": "sh:IRI",
+            "path": "did:keyAgreement"
```
```suggestion
            "path": "sec:keyAgreement"
```

msporny commented on vocabs/DID-core-shape-v1.jsonld:
```diff
@@ -0,0 +1,228 @@
+        {
+            "@id": "_:b16",
+            "nodeKind": "sh:IRI",
+            "path": "did:assertionMethod"
```
```suggestion
            "path": "sec:assertionMethod"
```

msporny commented on vocabs/DID-core-shape-v1.jsonld:
```diff
@@ -0,0 +1,228 @@
+            "@id": "_:b19",
+            "sh:maxCount": 1,
+            "sh:minCount": 1,
+            "path": "did:publicKeyJwk"
```
```suggestion
            "path": "sec:publicKeyJwk"
```

msporny commented on vocabs/DID-core-shape-v1.jsonld:
```diff
@@ -0,0 +1,228 @@
+            "@id": "_:b2",
+            "datatype": "xsd:string",
+            "nodeKind": "sh:Literal",
+            "path": "did:publicKeyBase58",
```
```suggestion
            "path": "sec:publicKeyBase58",
```

msporny commented on vocabs/DID-core-shape-v1.jsonld:
```diff
@@ -0,0 +1,228 @@
+            "@id": "_:b20",
+            "sh:maxCount": 1,
+            "sh:minCount": 1,
+            "path": "did:publicKeyBase58"
```
```suggestion
            "path": "sec:publicKeyBase58"
```

msporny commented on vocabs/DID-core-shape-v1.jsonld:
```diff
@@ -0,0 +1,228 @@
+            ],
+            "targetClass": "did:VerificationMethod",
+            "targetObjectsOf": [
+                "did:assertionMethod",
+                "did:capabilityDelegation",
+                "did:capabilityInvocation",
+                "did:authentication",
+                "did:keyAgreement",
+                "did:verificationMethod"
```
```suggestion
                "sec:assertionMethod",
                "sec:capabilityDelegation",
                "sec:capabilityInvocation",
                "sec:authentication",
                "sec:keyAgreement",
                "sec:verificationMethod"
```

msporny commented on vocabs/DID-core-shape-v1.jsonld:
```diff
@@ -0,0 +1,228 @@
+                "@type": "sh:NodeShape",
+                "property": [
+                    {
+                        "path": "did:publicKeyBase58",
```
```suggestion
                        "path": "sec:publicKeyBase58",
```

msporny commented on vocabs/DID-core-shape-v1.jsonld:
```diff
@@ -0,0 +1,228 @@
+                        "sh:minCount" : 1
+                    },
+                    {
+                        "path": "did:publicKeyJwk",
```
```suggestion
                        "path": "sec:publicKeyJwk",
```

msporny commented on vocabs/DID-core-shape-v1.jsonld:
```diff
@@ -0,0 +1,228 @@
+            "@id": "_:b6",
+            "datatype": "rdf:JSON",
+            "nodeKind": "sh:Literal",
+            "path": "did:publicKeyJwk",
```
```suggestion
            "path": "sec:publicKeyJwk",
```

msporny commented on vocabs/DID-core-shape-v1.jsonld:
```diff
@@ -0,0 +1,228 @@
+        {
+            "@id": "_:b7",
+            "sh:minCount": 1,
+            "path": "did:controller",
```
```suggestion
            "path": "sec:controller",
```

msporny commented on vocabs/DID-core-shape-v1.shex:
```diff
@@ -0,0 +1,39 @@
+# 
+
+#
+PREFIX did:  <https://www.w3.org/ns/did/vocab/v1#>
```
```suggestion
PREFIX did:  <https://www.w3.org/ns/did#>
PREFIX sec:  <https://w3id.org/security#>
```

iherman replied:
@msporny 

> @iherman what happens if you put a `did.var` file in the `/ns/` directory to purely direct it to the HTML description?

tried it, ain't working. It seems that the MultiViews feature ignores the `did.var` file if a `did` directory name is also present. This is exactly the problem. I tried to look around the web for solutions, did not find anything working. There were some proposals to try to switch MultiViews off, but this is not something I should do in the `/ns` directory, because it may have consequences on other things in the same directory. Ie, that is a big no-no.

msporny commented on vocabs/DID-core-shape-v1.shex:
```diff
@@ -0,0 +1,39 @@
+_:ControllerShape IRI /^did:[a-z0-9]+:[a-zA-Z0-9.-_:]+/
+
+_:VerificationMethodShape IRI {
+    did:controller @_:ControllerShape+ ;
```
```suggestion
    sec:controller @_:ControllerShape+ ;
```

msporny commented on vocabs/DID-core-shape-v1.shex:
```diff
@@ -0,0 +1,39 @@
+_:VerificationMethodShape IRI {
+    did:controller @_:ControllerShape+ ;
+    rdf:type IRI+ ;
+    ( did:publicKeyJwk rdf:JSON {0,1} | did:publicKeyBase58 LITERAL {0,1} ) {0,1};
```
```suggestion
    ( sec:publicKeyJwk rdf:JSON {0,1} | sec:publicKeyBase58 LITERAL {0,1} ) {0,1};
```

msporny commented on vocabs/DID-core-shape-v1.shex:
```diff
@@ -0,0 +1,39 @@
+# The id is not a blank node, because that is the node referred to from the outside world
+<SubjectShape> IRI /^did:[a-z0-9]+:[a-zA-Z0-9.-_:]+/ {
+    did:alsoKnownAs IRI* ;
+    did:controller @_:ControllerShape* ;
+    did:verificationMethod @_:VerificationMethodShape* ;
+    did:authentication @_:VerificationMethodShape* ;
+    did:assertionMethod @_:VerificationMethodShape* ;
+    did:keyAgreement @_:VerificationMethodShape* ;
+    did:capabilityDelegation @_:VerificationMethodShape* ;
+    did:capabilityInvocation @_:VerificationMethodShape* ;
```
```suggestion
    did:controller @_:ControllerShape* ;
    sec:verificationMethod @_:VerificationMethodShape* ;
    sec:authentication @_:VerificationMethodShape* ;
    sec:assertionMethod @_:VerificationMethodShape* ;
    sec:keyAgreement @_:VerificationMethodShape* ;
    sec:capabilityDelegation @_:VerificationMethodShape* ;
    sec:capabilityInvocation @_:VerificationMethodShape* ;
```

msporny commented on vocabs/DID-core-shape-v1.ttl:
```diff
@@ -0,0 +1,125 @@
+#  pyshacl -s DID-core-shape-v1.ttl -e DID-core-v1.ttl -f human shape_test.ttl
+#
+
+@prefix did:  <https://www.w3.org/ns/did/vocab/v1#> .
```
```suggestion
@prefix did:  <https://www.w3.org/ns/did/v1#> .
@prefix sec:  <https://w3id.org/security#> .
```

msporny replied:
You are correct, that is wrong... I was in the middle of lots of edits and messed this one up. Fixed.

msporny commented on vocabs/DID-core-shape-v1.ttl:
```diff
@@ -0,0 +1,125 @@
+    sh:targetClass did:DIDSubject ;
+    sh:targetSubjectsOf
+        did:alsoKnownAs,
+        did:verificationMethod,
+        did:authentication,
+        did:assertionMethod,
+        did:keyAgreement,
+        did:capabilityDelegation,
+        did:capabilityInvocation,
```
```suggestion
        sec:verificationMethod,
        sec:authentication,
        sec:assertionMethod,
        sec:keyAgreement,
        sec:capabilityDelegation,
        sec:capabilityInvocation,
```

msporny commented on vocabs/DID-core-shape-v1.ttl:
```diff
@@ -0,0 +1,125 @@
+            sh:path did:controller ;
+            sh:pattern "^did:[a-z0-9]+:[a-zA-Z0-9.-_:]+" ;
+        ],[
+            sh:path did:verificationMethod ;
```
```suggestion
            sh:path sec:verificationMethod ;
```

msporny commented on vocabs/DID-core-shape-v1.ttl:
```diff
@@ -0,0 +1,125 @@
+            sh:path did:verificationMethod ;
+            sh:nodeKind sh:IRI ;
+        ],[
+            sh:path did:authentication ;
```
```suggestion
            sh:path sec:authentication ;
```

msporny commented on vocabs/DID-core-shape-v1.ttl:
```diff
@@ -0,0 +1,125 @@
+            sh:path did:authentication ;
+            sh:nodeKind sh:IRI ;
+        ],[
+            sh:path did:assertionMethod ;
```
```suggestion
            sh:path sec:assertionMethod ;
```

msporny commented on vocabs/DID-core-shape-v1.ttl:
```diff
@@ -0,0 +1,125 @@
+            sh:path did:assertionMethod ;
+            sh:nodeKind sh:IRI ;
+        ],[
+            sh:path did:keyAgreement ;
```
```suggestion
            sh:path sec:keyAgreement ;
```

msporny commented on vocabs/DID-core-shape-v1.ttl:
```diff
@@ -0,0 +1,125 @@
+            sh:path did:keyAgreement ;
+            sh:nodeKind sh:IRI ;
+        ],[
+            sh:path did:capabilityDelegation ;
```
```suggestion
            sh:path sec:capabilityDelegation ;
```

msporny commented on vocabs/DID-core-shape-v1.ttl:
```diff
@@ -0,0 +1,125 @@
+            sh:path did:capabilityDelegation ;
+            sh:nodeKind sh:IRI ;
+        ],[
+            sh:path did:capabilityInvocation ;
```
```suggestion
            sh:path sec:capabilityInvocation ;
```

msporny commented on vocabs/DID-core-shape-v1.ttl:
```diff
@@ -0,0 +1,125 @@
+    # Generic shape for verification method, putting constraints on the allowed properties
+    sh:targetClass did:VerificationMethod ;
+    sh:targetObjectsOf
+        did:verificationMethod,
+        did:authentication,
+        did:assertionMethod,
+        did:keyAgreement,
+        did:capabilityDelegation,
+        did:capabilityInvocation ;
```
```suggestion
        sec:verificationMethod,
        sec:authentication,
        sec:assertionMethod,
        sec:keyAgreement,
        sec:capabilityDelegation,
        sec:capabilityInvocation ;
```

msporny commented on vocabs/DID-core-shape-v1.ttl:
```diff
@@ -0,0 +1,125 @@
+        did:capabilityInvocation ;
+    sh:property
+        [
+            sh:path did:publicKeyJwk ;
```
```suggestion
            sh:path sec:publicKeyJwk ;
```

msporny commented on vocabs/DID-core-shape-v1.ttl:
```diff
@@ -0,0 +1,125 @@
+            sh:datatype rdf:JSON ;
+            sh:maxCount 1;
+        ],[
+            sh:path did:publicKeyBase58 ;
```
```suggestion
            sh:path sec:publicKeyBase58 ;
```

msporny commented on vocabs/DID-core-shape-v1.ttl:
```diff
@@ -0,0 +1,125 @@
+            sh:datatype xsd:string ;
+            sh:maxCount 1;
+        ],[
+            sh:path did:controller ;
```
```suggestion
            sh:path sec:controller ;
```

msporny commented on vocabs/DID-core-shape-v1.ttl:
```diff
@@ -0,0 +1,125 @@
+            a sh:NodeShape ;
+            sh:property
+                [
+                    sh:path did:publicKeyBase58 ;
```
```suggestion
                    sh:path sec:publicKeyBase58 ;
```

msporny commented on vocabs/DID-core-shape-v1.ttl:
```diff
@@ -0,0 +1,125 @@
+                    sh:path did:publicKeyBase58 ;
+                    sh:minCount 1 ;
+                ],[
+                    sh:path did:publicKeyJwk ;
```
```suggestion
                    sh:path sec:publicKeyJwk ;
```

msporny commented on vocabs/DID-core-v1.jsonld:
```diff
@@ -0,0 +1,159 @@
+  },
+  "@graph": [
+    {
+      "@id": "https://www.w3.org/ns/did/vocab/v1#",
```
```suggestion
      "@id": "https://www.w3.org/ns/did#",
```

msporny commented on vocabs/DID-core-v1.jsonld:
```diff
@@ -0,0 +1,159 @@
+      "sameAs": "as:alsoKnownAs"
+    },
+    {
+      "@id": "did:assertionMethod",
```
```suggestion
      "@id": "sec:assertionMethod",
```

msporny commented on vocabs/DID-core-v1.jsonld:
```diff
@@ -0,0 +1,159 @@
+      "@id": "did:VerificationMethod",
+      "@type": "rdfs:Class",
+      "comment": "Superclass for all specific verification Methods. It is an 'abstract' class, never used by itself.",
+      "isDefinedBy": "https://www.w3.org/TR/did-core/#verification-methods",
```
```suggestion
      "isDefinedBy": "https://w3id.org/security#VerificationMethod",
```

msporny commented on vocabs/DID-core-v1.jsonld:
```diff
@@ -0,0 +1,159 @@
+      "@id": "did:assertionMethod",
+      "@type": "rdf:Property",
+      "comment": "The assertionMethod property is used to express a verification relationship which indicates that a verification method can be used to verify a proof that a statement was asserted on behalf of the DID subject.",
+      "isDefinedBy": "https://www.w3.org/TR/did-core/#assertionmethod",
```
```suggestion
      "isDefinedBy": "https://w3id.org/security#assertionMethod",
```

iherman replied:
I cannot really agree with this (and similar). `assertionMethod` is defined by the normative DID Specification, and not by the Security Vocabulary CG draft report (which, actually, labels this is 'unstable'). Surely the Rec has a priority over the draft report...

OR13 replied:
I am so glad this is finally being said.... when it comes to signatures, they change if this changes... which is why w3id.org was created, so that it could redirect to the appropriate normative definition....

https://w3id.org/security -> https://www.w3.org/TR/did-core/ 

^ this won't work because did core does not define all the terms that security needs and they are identified by fragment, so we cannot redirect there for 1/2 and somewhere else for the other, (like we could for schema.org definitions).

msporny replied:
> Surely the Rec has a priority over the draft report.

We could say that... and if we say that, we will invalidate all Linked Data Security implementations, all Verifiable Credentials issued to date, and all DID Method testnets and production deployments. :)

As Orie mentions above -- we can't redefine what the ecosystem is doing today w/o asking all implementers to make breaking changes to all of their implementations (and the only gain there will be is design purity, but at the cost of breaking the already deployed ecosystem).

In short, I think it's going to be a very difficult thing to change the URL at this point.

iherman replied:
??? Are we talking about the same thing?

All the terms and URI-s are now in the original, `sec:` vocabulary. The _only_ thing I am questioning here is whether the `rdfs:isDefinedBy` in the RDFS Vocabulary definition should point at the DID Core spec or the Security Draft. Surely none of the signature methods use this `rdfs:isDefinedBy`, which did not even exist (for DID) a month ago!

msporny commented on vocabs/DID-core-v1.jsonld:
```diff
@@ -0,0 +1,159 @@
+      "@type": "rdf:Property",
+      "comment": "The assertionMethod property is used to express a verification relationship which indicates that a verification method can be used to verify a proof that a statement was asserted on behalf of the DID subject.",
+      "isDefinedBy": "https://www.w3.org/TR/did-core/#assertionmethod",
+      "range": "did:VerificationMethod",
+      "sameAs": "sec:assertionMethod"
```
Now I'm confused... I don't think we should be duplicating classes w/ owl:sameAs... I do get the issue... I'm going to stop making suggestions from this point forward as we're probably not on the same page and need to get there... perhaps an interactive editing session might help.

iherman replied:
Yes, we should have a separate discussion on all this.

If we roll back the 'owl:sameAs' statements then:

- the DID vocabulary (RDFS, etc) in its own namespace consist of, from the top of my head, two terms, related to services. That is it.
- we could of course keep the current vocabulary files and keep everything to `sec:` but that looks quite strange: there is a namespace _for another_ vocabulary, whose proper definitions are defined in the namespace document for the `did:` vocabulary; moreover, the `sec:` vocabulary does not really have a proper definition. Messy.

We can do this of course, but I think it is ugly, and is really close to an anti-pattern for vocabulary definition. (I will hold my nose and not formally object, just to avoid confusion on that.)

msporny replied:
I think it's safe to say that none of us like where we are right now and ideally, we'd be depending on the Linked Data Security Vocabularies in a W3C namespace... but alas, we were blocked from starting there until there was demand for this stuff... and now that there is demand, we're having to do ugly things like this.

We'll just need to pick which ugly thing we think we can fix in time.

iherman replied:
@msporny o.k. I will roll back the `sameAs` statements, hold my nose, and get a cold beer from you when we meet next time personally (yeah, I know :-(

But I am still in the dark about the URL-s we should adopt for all these. Please, try to find some time going through the (tons of) comments about it because there are too many different branches now. See:

- https://github.com/w3c/did-spec-registries/pull/183#issuecomment-759232149
- https://github.com/w3c/did-spec-registries/pull/183#issuecomment-765561004
- https://github.com/w3c/did-spec-registries/pull/183#issuecomment-766094311
- https://github.com/w3c/did-spec-registries/pull/183#issuecomment-766099819
- https://github.com/w3c/did-spec-registries/pull/183#issuecomment-766331323
- https://github.com/w3c/did-spec-registries/pull/183#issuecomment-770287530
- https://github.com/w3c/did-spec-registries/pull/183#issuecomment-770346191
- https://github.com/w3c/did-spec-registries/pull/183#issuecomment-770981036

There is now a total confusion :-)

msporny reviewed (changes requested):
Overall, thank you! -- almost all of this is good. I have two major concerns:

1. The URLs we're using for the human and machine-readable vocabularies.
2. The class shadowing that's going on here... did:authentication -> sec:authentication -- it's confusing to those that know what's going on, and I expect it to be all but impossible for those that don't know what's going on to understand it.

OR13 commented:
> 2\. The class shadowing that's going on here... did:authentication -> sec:authentication -- it's confusing to those that know what's going on, and I expect it to be all but impossible for those that don't know what's going on to understand it.

AGREED! this has been bothering me since I joined the WG :)

iherman commented:
I did some experimentation on the issues in https://github.com/w3c/did-spec-registries/pull/183#issuecomment-770346191 and my suspicion was correct. Doing

- `https://www.w3.org/ns/did#` for the vocabulary
- `https://www.w3.org/ns/did/v1` for the context file
- `https://www.w3.org/ns/did/shacl` for the shacl file
- `https://www.w3.org/ns/did/shex` for the shex file

all with proper content negotiations seems to be problematic. The reason is that setting up a proper conneg for `../did` and `../did/something` clashes with the way apache works. Indeed, if it gets `https://www.w3.org/ns/did` and there is no real file, like `did.jsonld` in `/ns` then it would add a '/' to the URL and search for a directory. This is the MultiViews feature of Apache.

I did experiment with a trick, essentially, adding 'index.var` into the `../ns/did/` folder and making use of what happens in MultiViews. I did this experimentation in my own area on our server. As a result:

- `https://www.w3.org/People/Ivan/ns/did/v1` works as expected (returning the context file) with conneg
- `https://www.w3.org/People/Ivan/ns/did/shacl` (and `shex`) works as expected with conneg

As for `https://www.w3.org/People/Ivan/ns/did#` if I put it into the browser bar then it returns the html reference (currently set to the registry spec for testing), `https://www.w3.org/People/Ivan/ns/did.ttl` works all right, as well as jsonld, but conneg does not work. If I do a:

```
curl -head --header "Accept: application/ld+json" https://www.w3.org/People/Ivan/ns/did
```

I get 

```
date: Mon, 01 Feb 2021 16:22:36 GMT
location: https://www.w3.org/People/Ivan/ns/did/
cache-control: max-age=21600
expires: Mon, 01 Feb 2021 22:22:36 GMT
content-type: text/html; charset=iso-8859-1
x-backend: ssl-mirrors
strict-transport-security: max-age=15552000; includeSubdomains; preload
content-security-policy: upgrade-insecure-requests
```

@rhiaro @msporny @dlongley if you can experiment with it, that would be fine.

---

I am now more and more convinced that the proper solution is:

- `https://www.w3.org/ns/did/v1` for the context file
- `https://www.w3.org/ns/did/vocab` for the vocabulary file, i.e., that should be the namespace URI for the did vocabulary.

Note that if we indeed keep everything in the `sec:` namespace, then we are talking about 3 classes and two properties altogether...

rhiaro commented:
@ivan what happens if you put a `did.var` file in the `/ns/` directory to purely direct it to the HTML description?

iherman commented:
@rhiaro

> @iherman what happens if you put a did.var file in the /ns/ directory to purely direct it to the HTML description?

tried it, ain't working. It seems that the MultiViews feature ignores the did.var file if a did directory name is also present. This is exactly the problem. I tried to look around the web for solutions, did not find anything working. There were some proposals to try to switch MultiViews off, but this is not something I should do in the /ns directory, because it may have consequences on other things in the same directory. Ie, that is a big no-no.

msporny commented:
> I am now more and more convinced that the proper solution is

Counter example, showing that what I'm suggesting is possible:

https://w3id.org/security
https://w3id.org/security#Ed25519Signature2020
https://w3id.org/security/v1

... and the Apache-compatible .htaccess file:

https://github.com/perma-id/w3id.org/blob/master/security/.htaccess

This is how w3id.org has worked for years with over 4,950+ commits and 400 contributors... it is possible. :)

iherman commented:
@msporny 

yes. But look at:

```
18:06 nstest> curl --header "Accept: application/ld+json" https://w3id.org/security 
<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<html><head>
<title>301 Moved Permanently</title>
</head><body>
<h1>Moved Permanently</h1>
<p>The document has moved <a href="https://w3id.org/security/">here</a>.</p>
<hr>
<address>Apache/2.4.29 (Ubuntu) Server at w3id.org Port 443</address>
</body></html>
```

this is what I have, see https://github.com/w3c/did-spec-registries/pull/183#issuecomment-770981036

But compare with 

```
18:08 nstest> curl --header "Accept: application/ld+json" https://www.w3.org/People/Ivan/ns/did/shacl
```

which 'simply' returns the json-ld file. Are we sure that all tools handle the type of redirection above without a glitch? This is my concern.

@dlongley @rhiaro you may want to test your tools. I remember @rhiaro was complaining about some setups at the beginning of the WG and she was happy to have the `.var` solution that works like charm...

OR13 commented:
> > I am now more and more convinced that the proper solution is
> 
> Counter example, showing that what I'm suggesting is possible:
> 
> https://w3id.org/security
> https://w3id.org/security#Ed25519Signature2020
> https://w3id.org/security/v1
> 
> ... and the Apache-compatible .htaccess file:
> 
> https://github.com/perma-id/w3id.org/blob/master/security/.htaccess
> 
> This is how w3id.org has worked for years with over 4,950+ commits and 400 contributors... it is possible. :)

I would really prefer we just do exactly the same thing in both places... 

However, such a change cannot be made "here" since this is not the repo for managing the `.htaccess` file.... this repo needs to host the files we intend to expose... 

We should start by agreeing with the URLS for github, and then agree to the URLs for W3C.... 

https://w3c.github.io/did-spec-registries/contexts/did-v1.jsonld
https://w3c.github.io/did-spec-registries/index.html

^ these are the only urls currently exposed by this repo, and they map to:

https://www.w3.org/ns/did/v1
https://www.w3.org/TR/did-spec-registries/

This PR adds new files... which will produce:

https://w3c.github.io/did-spec-registries/vocabs/DID-core-v1.jsonld
https://w3c.github.io/did-spec-registries/vocabs/DID-core-v1.rdf
https://w3c.github.io/did-spec-registries/vocabs/DID-core-v1.ttl 
https://w3c.github.io/did-spec-registries/vocabs/DID-core-shape-v1.jsonld
https://w3c.github.io/did-spec-registries/vocabs/DID-core-shape-v1.shex
https://w3c.github.io/did-spec-registries/vocabs/DID-core-shape-v1.ttl
https://w3c.github.io/did-spec-registries/vocabs/readme.md (this will be rendered as html by github)

Lets focus this PR on changes to this repo, and use issues to track requests for redirects that map to this repo, for example:

- changes to w3.org or w3id.org

personally I would prefer we follow the existing convention established by:

https://w3c.github.io/did-spec-registries/contexts/did-v1.jsonld
https://w3c.github.io/did-spec-registries/index.html

and instead see the following after this PR is merged:

https://w3c.github.io/did-spec-registries/vocabs/v1/index.html

https://w3c.github.io/did-spec-registries/vocabs/v1/did-v1.jsonld
https://w3c.github.io/did-spec-registries/vocabs/v1/did-v1.rdf
https://w3c.github.io/did-spec-registries/vocabs/v1/did-v1.ttl

https://w3c.github.io/did-spec-registries/vocabs/v1/did-shape-v1.jsonld
https://w3c.github.io/did-spec-registries/vocabs/v1/did-shape-v1.shex
https://w3c.github.io/did-spec-registries/vocabs/v1/did-shape-v1.ttl

It feels like context maybe belongs under vocab.... but I would prefer to tackle that change in a seperate PR.

This PR should be about what are the files we are adding, where are we adding them, and what is in them.

OR13 commented:
We are having a related conversation here: https://github.com/w3c-ccg/traceability-vocab/issues/65#issuecomment-771089106

iherman commented:
> We should start by agreeing with the URLS for github, and then agree to the URLs for W3C....

Unfortunately, this is not really doable. All the files refer to the DID vocabulary namespace URL which therefore should be correct.

iherman commented:
@msporny, https://github.com/w3c/did-spec-registries/pull/183/commits/4f23984ea2ae289aa3fe91af87a614ff5a8932e9 took care of your comments (leaving the URI decision open):

- removed the `owl:sameAs` statements, meaning that the vocabulary puts information on some of the `sec` terms
- formatted the JSON-LD output of the Turtle->JSON-LD converters

OR13 commented:
pending changes from @iherman

iherman commented:
I have made the commit https://github.com/w3c/did-spec-registries/pull/183/commits/cd8c4892cd0478ccd21da97320594cab89bd461c along the lines of what was discussed with @rhiaro, @msporny, @OR13. The changes are as follows:

- all the files are regrouped in the `vocabs/v1` directory. There is a [readme.md](https://raw.githack.com/w3c/did-spec-registries/cd8c4892cd0478ccd21da97320594cab89bd461c/vocabs/v1/readme.md) file that describes the details of the content. Note that this directory will be the final place, within the repository, for the JSON-LD context file.
- the URL-s to be used (eventually, when the redirections are set up on `https://www.w3.org/ns/`):
    - The context file's URI is (as before) `https://www.w3.org/ns/did/v1`.
    - The official vocabulary namespace URI is `https://www.w3.org/ns/did`. This URI is conneg-d for a HTML, Turtle, json-ld, and rdf/xml version. 
    - The official URI for the shacl file is `https://www.w3.org/ns/shacl`. This URI is conneg-d for Turtle or json-ld
    - The official URI for the shex file `https://www.w3.org/ns/shex`. This URI refers to the shex syntax only.

@rhiaro, @msporny, @OR13, some minor changes compared to the discussion we had yesterday:

- The [`vocab.html`](https://raw.githack.com/w3c/did-spec-registries/cd8c4892cd0478ccd21da97320594cab89bd461c/vocabs/v1/vocab.html) required quite some changes after all. This is a vocabulary file only (not a human facing version of the context file), I had to add some explanations due to the unusual setup v.à.v. the security vocabulary, add reference to the classes (this is necessary for the vocabulary file), etc. Please review. (I hope it is all right that I have added my name as a co-editor; people should come to me, too, if there are some changes to be done.)
- In view of the previous point I have copied, but left unchanged, the `context.html` file, ie, a human facing version to the context file. I am not sure what should happen with that one.
- In contrast to what I said yesterday, I have added the .htaccess files into the repo as well. It is good if it is easy to review it and keep it in the open.

We are getting there... :-)

msporny reviewed (approved):
I have concerns about the .../did/shacl and .../did/shex redirects... think they should be .../did/shacl/v1 and .../did/shex/v1 -- but we can debate/fix that later. This PR has gotten large and it is clearly an improvement over what we have today, so let's please merge it and work from there.

Thank you, @iherman, for putting all this together! A herculean task.
